File: scripts/okf-index-lookup.py
#!/usr/bin/env python3
"""okf-index-lookup — grep knowledge/index.md for a term and emit matching concept-file paths.

Standalone. No Headroom dependency. Stdlib only.

Usage:
    python scripts/okf-index-lookup.py <term> [--root <repo-root>] [--limit N]

Behaviour:
- Loads knowledge/index.md relative to <repo-root> (default: parent of this script's dir).
- Case-insensitive substring match against each line.
- Extracts every markdown link target on matching lines: `[label](./path/to/file.md)`.
- Filters to `.md` paths (drops anchors, external URLs).
- Resolves each path relative to knowledge/index.md's directory and prints absolute paths
  that exist on disk. Deduped, in first-seen order.
- Exit 0 if at least one path printed; 1 otherwise. Empty stdout when no match.
"""
from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

LINK_RE = re.compile(r"\[[^\]]+\]\(([^)]+)\)")


def find_repo_root(start: Path) -> Path:
    # Default: parent of scripts/ — i.e. the repo root that owns knowledge/index.md.
    return start.resolve().parent.parent


def lookup(term: str, repo_root: Path, limit: int | None) -> list[Path]:
    index_path = repo_root / "knowledge" / "index.md"
    if not index_path.is_file():
        print(f"error: index not found at {index_path}", file=sys.stderr)
        sys.exit(2)

    needle = term.lower()
    index_dir = index_path.parent
    seen: set[Path] = set()
    out: list[Path] = []

    for raw in index_path.read_text(encoding="utf-8", errors="replace").splitlines():
        if needle not in raw.lower():
            continue
        for target in LINK_RE.findall(raw):
            target = target.split("#", 1)[0].strip()
            if not target.endswith(".md"):
                continue
            if target.startswith(("http://", "https://", "mailto:")):
                continue
            resolved = (index_dir / target).resolve()
            try:
                resolved.relative_to(repo_root.resolve())
            except ValueError:
                continue
            if not resolved.is_file():
                continue
            if resolved in seen:
                continue
            seen.add(resolved)
            out.append(resolved)
            if limit and len(out) >= limit:
                return out
    return out


def main() -> int:
    p = argparse.ArgumentParser(description="Grep knowledge/index.md for a term.")
    p.add_argument("term", help="Case-insensitive substring to search for.")
    p.add_argument(
        "--root",
        type=Path,
        default=None,
        help="Repo root (default: parent of this script's dir).",
    )
    p.add_argument("--limit", type=int, default=None, help="Cap on matches printed.")
    args = p.parse_args()

    root = args.root if args.root else find_repo_root(Path(__file__))
    paths = lookup(args.term, root, args.limit)
    for path in paths:
        print(path)
    return 0 if paths else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/okf-index-lookup.ts
#!/usr/bin/env -S npx tsx
// okf-index-lookup — grep knowledge/index.md for a term and emit matching concept-file paths.
// TypeScript port of okf-index-lookup.py. Stdlib-only.

import * as fs from "node:fs";
import * as path from "node:path";
import * as process from "node:process";
import { fileURLToPath } from "node:url";

const LINK_RE = /\[[^\]]+\]\(([^)]+)\)/g;

function findRepoRoot(start: string): string {
  // Default: parent of scripts/ — i.e. the repo root that owns knowledge/index.md.
  return path.dirname(path.dirname(path.resolve(start)));
}

function isSubPath(child: string, parent: string): boolean {
  const rel = path.relative(parent, child);
  return !rel.startsWith("..") && !path.isAbsolute(rel);
}

function lookup(term: string, repoRoot: string, limit: number | null): string[] {
  const indexPath = path.join(repoRoot, "knowledge", "index.md");
  if (!fs.existsSync(indexPath) || !fs.statSync(indexPath).isFile()) {
    process.stderr.write(`error: index not found at ${indexPath}\n`);
    process.exit(2);
  }

  const needle = term.toLowerCase();
  const indexDir = path.dirname(indexPath);
  const repoResolved = path.resolve(repoRoot);
  const seen = new Set<string>();
  const out: string[] = [];

  const content = fs.readFileSync(indexPath, "utf-8");
  const lines = content.split(/\r?\n/);

  for (const raw of lines) {
    if (!raw.toLowerCase().includes(needle)) continue;
    LINK_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = LINK_RE.exec(raw)) !== null) {
      let target = m[1].split("#", 1)[0].trim();
      if (!target.endsWith(".md")) continue;
      if (
        target.startsWith("http://") ||
        target.startsWith("https://") ||
        target.startsWith("mailto:")
      )
        continue;
      const resolved = path.resolve(path.join(indexDir, target));
      if (!isSubPath(resolved, repoResolved) && resolved !== repoResolved) continue;
      if (!fs.existsSync(resolved) || !fs.statSync(resolved).isFile()) continue;
      if (seen.has(resolved)) continue;
      seen.add(resolved);
      out.push(resolved);
      if (limit && out.length >= limit) return out;
    }
  }
  return out;
}

function main(): number {
  const argv = process.argv.slice(2);
  let term: string | undefined;
  let root: string | undefined;
  let limit: number | null = null;

  let i = 0;
  while (i < argv.length) {
    const a = argv[i];
    if (a === "--root") {
      root = argv[++i];
    } else if (a === "--limit") {
      limit = parseInt(argv[++i], 10);
    } else if (a === "-h" || a === "--help") {
      process.stdout.write("Usage: okf-index-lookup <term> [--root DIR] [--limit N]\n");
      return 0;
    } else if (a.startsWith("--")) {
      process.stderr.write(`error: unknown flag ${a}\n`);
      return 2;
    } else if (term === undefined) {
      term = a;
    }
    i++;
  }

  if (term === undefined) {
    process.stderr.write("error: term required\n");
    return 2;
  }

  const scriptFile = fileURLToPath(import.meta.url);
  const repoRoot = root ? path.resolve(root) : findRepoRoot(scriptFile);
  const paths = lookup(term, repoRoot, limit);
  for (const p of paths) process.stdout.write(p + "\n");
  return paths.length > 0 ? 0 : 1;
}

process.exit(main());
